bothan-api/server-cli/src/commands/query.rs

```diff
@@ -0,0 +1,359 @@
+    opts: bothan_htx::WorkerOpts,
+    query_ids: &[String],
+    timeout: T,
+) -> anyhow::Result<()> {
+    let connector = Arc::new(bothan_htx::api::WebSocketConnector::new(opts.url));
+    let asset_infos = query_websocket(connector, dedup(query_ids), timeout.into()).await?;
+
+    display_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_kraken<T: Into<Duration>>(
+    opts: bothan_kraken::WorkerOpts,
+    query_ids: &[String],
+    timeout: T,
+) -> anyhow::Result<()> {
+    let connector = Arc::new(bothan_kraken::api::WebSocketConnector::new(opts.url));
+    let asset_infos = query_websocket(connector, dedup(query_ids), timeout.into()).await?;
+
+    display_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_okx<T: Into<Duration>>(
+    opts: bothan_okx::WorkerOpts,
+    query_ids: &[String],
+    timeout: T,
+) -> anyhow::Result<()> {
+    let connector = Arc::new(bothan_okx::api::WebSocketConnector::new(opts.url));
+    let asset_infos = query_websocket(connector, dedup(query_ids), timeout.into()).await?;
+
+    display_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_websocket_with_max_sub<C, P, E1, E2>(
+    connector: Arc<C>,
+    ids: Vec<String>,
+    max_subscription_per_connection: usize,
+    timeout: Duration,
+) -> anyhow::Result<Vec<AssetInfo>>
+where
+    E1: Error + Send + Sync + 'static,
+    E2: Error + Send + Sync + 'static,
+    P: WebSocketAssetInfoProvider<SubscriptionError = E1, PollingError = E2>,
+    C: AssetInfoProviderConnector<Provider = P, Error = E1>,
+{
+    let tasks = FuturesUnordered::new();
+
+    for chunk in &ids.into_iter().chunks(max_subscription_per_connection) {
+        let chunk_ids = chunk.collect();
+        let cloned_connector = connector.clone();
+
+        tasks.push(async move { query_websocket(cloned_connector, chunk_ids, timeout).await });
+    }
+
+    let asset_infos = tasks
+        .try_collect::<Vec<Vec<AssetInfo>>>()
+        .await?
+        .into_iter()
+        .flatten()
+        .collect();
+
+    Ok(asset_infos)
+}
+
+async fn query_websocket<E1, E2, P, C>(
+    connector: Arc<C>,
+    ids: Vec<String>,
+    timeout_interval: Duration,
+) -> anyhow::Result<Vec<AssetInfo>>
+where
+    E1: Error + Send + Sync + 'static,
+    E2: Error + Send + Sync + 'static,
+    P: WebSocketAssetInfoProvider<SubscriptionError = E1, PollingError = E2>,
+    C: AssetInfoProviderConnector<Provider = P, Error = E1>,
+{
+    let mut provider = connector.connect().await?;
+    provider.subscribe(&ids).await?;
+
+    let mut asset_infos: HashMap<String, AssetInfo> = HashMap::with_capacity(ids.len());
+
+    timeout(timeout_interval, async {
+        while asset_infos.len() < ids.len() {
+            let data = provider.next().await?;
+
+            if let Ok(Data::AssetInfo(infos)) = data {
+                for info in infos {
+                    asset_infos.insert(info.id.clone(), info);
+                }
+            }
+        }
+        Some(())
+    })
+    .await?
```

Comment:
does this mean it can't show partial result?  for example, if some signal make it timeout, it won't show any result for other signals, right?